### ISSUE TITLE
Keychain/Key manager/Identity manager

### ISSUE THREAD

**Issue by dennismckinnon** [eris-ltd/decerver#78]
I've been thinking about this for a while. Clearly keys being leaked to malicious modules is probably the single greatest security risk we face. But at the same time having an easy and simple method by which to   manage keys across modules and group them into units seems to be a key requirement. (har har)

My thinking on this as follows.

Keys should be stored in a central location and they should be categorized and formatted such that a key stores the private, public, type, maybe a few other things as well. 

The idea is that the key need not be tied to a particular module. Not sure this is necessary or helpful but it seems like it might be.

Thats storage. In my head we have the .keys folder with subfolders for each type of key. each key would be stored in a json file with some standard files and then some optional fields that are module specific so we only need to come up with the bare minimum required for a key. (for example private puclic might even be unnecessary. instead "Secret" is the mandatory field and public key public address etc are optional. under the .keys folder we also should have an Identities folder. 

The identity has a name for easy reference and fields specifying the key that should be used (either on a per type or per module basis.

For using it, module wrappers should have key gen and key submit  key data functions something like that which specifies what the module is expecting. when loading a dapp you should choose an identity which will specify these choices for keys and submit them into the module. This action should require (at least the first time) authorization for the dapp to use that identity and should be run in the decerver itself. 

If your identity does not contain the requirements for a particular module. then the option to create a key or to add an existing key to the identity should be presented.

I think this balances the control and the ease of use. happy for input @androlo @compleatang @ebuchman

**Comment by androlo** [comment]
I like the idea of a key manager, and a module wrapper for it. We could use it for all the decerver stuff and some of the core modules. I think we should not make it a general thing for modules, to avoid a potential security and maintenance nightmare.

**Comment by dennismckinnon** [comment]
I'm not sure it should be a module. Requests to the key manager should have to pass through the decerver directly so that we can enforce certain security restrictions on it. Thats sort of why I thought we should also enforce a standard on module-wrappers to follow certain requirements in order to use keys from the key manager.

The thinking is that a person should have always authorized the module to access the keys and not give that control to a dapp. Ideally the dapp never would have access to the key directly as this would be most susceptible to malicious actions.

Just my thinking on the matter we can discuss more tomorrow morning.

**Comment by androlo** [comment]
The way to provide the key manager to the modules directly is easy - just add it to the set of core components that is exposed to the module during registration (like the js runtime manager, event processor etc.). We would not have to worry about authenticating modules, at least not until we have dynamic loading, so it should be fine. 

Personally don't think core decerver should have any code for doing crypto, whether it's signing messages, generating keys, or even storing them. I'd like to keep that type of stuff at arms length. Naming keys and grouping them, and help passing them around would be very useful though I agree. I think it might even be necessary.

**Comment by dennismckinnon** [comment]
current plan as from meeting earlier today, Put the central key store on hold for the moment until we are more comfortable with the security aspects which may arise. Instead improve the wrappers to include ability to choose keys/keyrings. As a temp solution only. Identity system I still think should be implemented at the decerver level so that when we are ready for the keystore system we have that framework in place. the identity files need only have module:keyreference pairs so during dapp start up an identity can be selected.

For now I think this will be sufficient. In the future for mobile capability as well as remote capability, and also convenience of being able to back up identities, we should still have the keystore on our radar.

**Comment by compleatang** [comment]
closing this for now. this issue is not finalized, but we will need to think how to implement given the new middlewear design pattern (e.g., decerver-node).
